src/lib/CofThy.ml

```diff
@@ -1,209 +1,233 @@
-    match irreducible_joins with
-    | [] -> true
-    | psis :: irreducible_joins ->
-      psis |> List.exists @@ fun psi ->
-      Option.fold ~none:false ~some:is_consistent @@
-      push_cofs {thy with irreducible_joins} [psi]
-end
-
-module Search :
-sig
-  (** Search all branches assuming more cofibrations. *)
-  val left_invert
-    : zero:'a
-    (** [zero] is the default value for vacuous cases. *)
-    -> seq:((cof -> 'a) -> cof list -> 'a)
-    (** [seq] is the sequencing operator. *)
-    -> fast_track:((unit -> 'a) -> (unit -> 'a) -> 'a)
-    (** If the first component returns a "good" result, then don't bother with the second. (???) *)
-    -> disj_thy
-    -> cof list
-    -> (alg_thy' -> 'a)
-    -> 'a
-end =
-struct
-  let left_invert' ~zero ~seq ~fast_track (thy : disj_thy') phis cont =
-    let rec go =
-      function
-      | None -> zero
-      | Some ({classes; true_vars; irreducible_joins} as thy) ->
-        fast_track (fun _ -> cont {classes; true_vars}) @@ fun _ ->
-        match irreducible_joins with
-        | [] -> cont {classes; true_vars}
-        | psis :: irreducible_joins when SearchHelper.is_consistent thy ->
-          psis |> seq @@ fun psi ->
-          go @@ SearchHelper.push_cofs {thy with irreducible_joins} [psi]
-        | _ ->
-          zero
+          test_eq thy' (r, s)
+        | Cof.Join phis ->
+          List.exists (test thy') phis
+        | Cof.Meet phis ->
+          List.for_all (test thy') phis
+      end
+    | Cof.Var v ->
+      test_var thy' v
+
+  let shrink_branches (thy' : t') branches : cached_branches =
+    (* stage 1.1: shrink branches *)
+    let go branch =
+      match normalize_branch thy' branch with
+      | `Inconsistent -> None
+      | `Consistent (thy', branch) -> Some (thy', branch)
     in
-    go @@ SearchHelper.push_cofs thy phis
+    List.filter_map go branches
+
+  let drop_useless_branches cached_branches : cached_branches =
+    let go_fwd acc (thy', branch) =
+      if Bwd.exists (fun (_, branch) -> test_branch thy' branch) acc then
+        acc
+      else
+        Snoc (acc, (thy', branch))
+    in
+    let cached_branches = List.fold_left go_fwd Emp cached_branches in
+    let go_bwd (thy', branch) acc =
+      if List.exists (fun (_, branch) -> test_branch thy' branch) acc then
+        acc
+      else
+        (thy', branch) :: acc
+    in
+    Bwd.fold_right go_bwd cached_branches []
 
-  let left_invert ~zero ~seq ~fast_track thy phis cont =
+  let split (thy : t) (cofs : cof list) : t list =
     match thy with
-    | `Inconsistent -> zero
-    | `Consistent thy -> left_invert' ~zero ~seq ~fast_track thy phis cont
+    | `Inconsistent -> []
+    | `Consistent thy' ->
+      match disect_cofibrations cofs with
+      | [] -> []
+      | [vars, []] when VarSet.is_empty vars -> [`Consistent thy']
+      | disected_cofs ->
+        let cached_branches =
+          drop_useless_branches @@
+          shrink_branches thy' disected_cofs in
```

Comment:
when a line is broken like this, I prefer `in` to appear on its own line below

Reply:
LOL the formatting was due to debugging... I will fix that.